chromaprint.py

```diff
@@ -44,7 +44,10 @@ def _load_library(name):
             return None
 
     try:
-        return ctypes.cdll.LoadLibrary(name)
+        if sys.platform == 'win32':
+            return ctypes.WinDLL(name, winmode=0)
+        else:
+            return ctypes.cdll.LoadLibrary(name)
```

Comment:
You mentioned that this didn't work for you without this change. Can you please include the error message so we can understand more about why this is the right fix?

Reply:
It gives me:
```python
----> 1 print(acoustid.compare_fingerprints(old_fingerprint, new_fingerprint))

File ~\AppData\Local\pypoetry\Cache\virtualenvs\packager-XchScg4l-py3.11\Lib\site-packages\acoustid.py:400, in compare_fingerprints(a, b)
    393 """Compare two fingerprints produced by `fingerprint_file`.
    394 
    395 :param a: A pair produced by `fingerprint_file`.
    396 :param b: A second such pair.
    397 :return:  similarity score [0,1]
    398 """
    399 if not have_chromaprint:
--> 400     raise ModuleNotFoundError("function needs chromaprint")
    402 # decompress fingerprints
    403 a = [int(x) for x in chromaprint.decode_fingerprint(a[1])[0]]

ModuleNotFoundError: function needs chromaprint
```
`have_chromaprint` is set to false, which means that `import chromaprint` raises `ImportError`.
`ImportError` is raised by `chromaprint.py` if `_load_library` does not load anything, if applying this change it will work, instead.

Reply:
Hmm; I wonder if we can get to the underlying reason here. Maybe you could try running `ctypes.WinDLL('libchromaprint.dll', winmode=0)` and `ctypes.cdll.LoadLibrary(('libchromaprint.dll')` in the Python REPL to see what happens?

Reply:
The documentation for `CDLL` and `WinDLL` may be clarifying (i.e., the difference seems to be in the calling convention that the library must use): https://docs.python.org/3.11/library/ctypes.html?highlight=ctypes#ctypes.CDLL

Reply:
I have indagated a little bit more and the problem is not about `CDLL` or `WinDLL` but, instead `winmode` set to something else than `None`.

Reply:
OK! I admit this makes me a little nervous, given the Python documentation's admonition about the flag:

> When omitted, the default is to use the flags that result in the most secure DLL load to avoiding issues such as DLL hijacking.

So it seems like setting `winmode=0` disables some security features. I am a total Windows novice so I can't say confidently how important those security features are, but it makes me worry that the implications may be deeper than they first appear. There's a list here:
https://learn.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-loadlibraryexa

Maybe it would be useful to figure out exactly what the problem is (which default security option is preventing loading on your system), and maybe expose this flag as user-configurable instead of defaulting it to `0`?

Reply:
Referring to [this](https://stackoverflow.com/questions/59330863/cant-import-dll-module-in-python), without setting `winmode` uses `nt._LOAD_LIBRARY_SEARCH_DEFAULT_DIRS` which has the same name on the page you linked, instead using `ctypes.DEFAULT_MODE`=0, it behaves identically to [LoadLibrary](https://learn.microsoft.com/en-us/windows/desktop/api/libloaderapi/nf-libloaderapi-loadlibrarya) as specified in the Remarks page. It seems that this behaviour is ok if the path to the dll is passed which is what we are doing, and also this means no hijacking.

Also referring to [this](http://winapi.freetechsecrets.com/win32/WIN32LoadLibraryEx.htm) website, the alternative mode is the right mode to use since the dll is not in the application folder, but in the path specified.
So a better alternative to `winmode=ctypes.DEFAULT_MODE` could be `winmode=8` which is `LOAD_WITH_ALTERED_SEARCH_PATH`.

So I think it is not a security issue.

Reply:
Got it; thank you so much for chasing this down! It looks like this was a legitimate regression in Python 3.8, as discussed in https://github.com/python/cpython/issues/86280. Great!!

It seems like the `winmode` parameter also exists on `CDLL`, so maybe we can get rid of the `if` here and just do:


```py
return ctypes.CDLL(name, winmode=ctypes.LOAD_WITH_ALTERED_SEARCH_PATH)
```

or something like that?

Reply:
`ctypes.LOAD_WITH_ALTERED_SEARCH_PATH` does not exist, I created a constant and wrote a comment.

(An alternative could be the use of `ctypes.DEFAULT_MODE` which is 0, that also works)